### PR TITLE
Replace supports_bounded_execution with supports_retract_batch

### PR DESCRIPTION
**Pull request by mustafasrepo**
# Which issue does this PR close?

Related to #5781 

# Rationale for this change
With the changes in the [#6671](https://github.com/apache/arrow-datafusion/pull/6671), `supports_retract_batch` method is introduced to the `Accumulator` trait. With the introduction of  `supports_retract_batch` method, `supports_bounded_execution` method is no longer necessary for the `AggregateExpr` trait. (Similar to the case we have move `supports_bounded_execution` trait from `BuiltinWindowFunctionExpr` to `PartitionEvalautor` trait.) 

This PR removes `supports_bounded_execution` method from `AggregateExpr` and moves its functionality to `supports_retract_batch` method in the Accumulator` trait for existing accumulators.

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/window/aggregate.rs [hunk 1]
```diff
@@ -155,8 +155,7 @@ impl WindowExpr for PlainAggregateWindowExpr {
     }
 
     fn uses_bounded_memory(&self) -> bool {
-        self.aggregate.supports_bounded_execution()
-            && !self.window_frame.end_bound.is_unbounded()
+        !self.window_frame.end_bound.is_unbounded()
```
Comment (mustafasrepo) [inline datafusion/physical-expr/src/window/aggregate.rs:158]: After thinking through this logic. Actually, as long as `end_bound` is not bounded (not `UNBOUNDED FOLLOWING` such as in the form `N FOLLOWING`). We can produce results without waiting for the whole data to come (If accumulator do not support `retract_batch` method. We wouldn't be able to run queries in the form `M PRECEDING and N FOLLOWING`, in this case we will give an error anyway.). Hence here we do not need to check for `self.aggregate.supports_bounded_execution()` (`acc.supports_retract_batch()` method with the new API.)

File: datafusion/physical-expr/src/window/sliding_aggregate.rs [hunk 2]
```diff
@@ -139,8 +139,7 @@ impl WindowExpr for SlidingAggregateWindowExpr {
     }
 
     fn uses_bounded_memory(&self) -> bool {
-        self.aggregate.supports_bounded_execution()
-            && !self.window_frame.end_bound.is_unbounded()
+        !self.window_frame.end_bound.is_unbounded()
```
Comment (mustafasrepo) [inline datafusion/physical-expr/src/window/sliding_aggregate.rs:142]: Similar to the case above.

File: datafusion/physical-expr/src/window/aggregate.rs (hunk 1 above)
Comment (alamb) [reply datafusion/physical-expr/src/window/aggregate.rs:158]: This makes sense to me

File: datafusion/physical-expr/src/aggregate/mod.rs [hunk 3]
```diff
@@ -96,12 +96,6 @@ pub trait AggregateExpr: Send + Sync + Debug + PartialEq<dyn Any> {
         false
     }
 
-    /// Specifies whether this aggregate function can run using bounded memory.
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/mod.rs:99]: THis is nice to have moved this logic entirely to the accumulator 👍

**Review by alamb — APPROVED** [review]
Thank you @mustafasrepo 🙏

**Pull request merged**